cohortextractor/tpp_backend.py

```diff
@@ -1992,24 +1987,32 @@ def __str__(self):
         return self.expression
 
 
-def codelist_to_sql_list(codelist):
+def codelist_to_list(codelist):
     if getattr(codelist, "has_categories", False):
-        return [quote(code) for (code, category) in codelist]
+        return [code for (code, category) in codelist]
     else:
-        return [quote(code) for code in codelist]
+        return list(codelist)
 
 
 def codelist_to_like_patterns(codelist, prefix="", suffix=""):
     patterns = []
-    for quoted_code in codelist_to_sql_list(codelist):
-        assert quoted_code[0] == "'"
-        assert quoted_code[-1] == "'"
-        patterns.append(f"'{prefix}{quoted_code[1:-1]}{suffix}'")
+    for code in codelist_to_list(codelist):
+        escaped_code = escape_like_query_fragment(code)
+        patterns.append(quote(f"{prefix}{escaped_code}{suffix}"))
     return patterns
 
 
+# Special characters from:
+# https://docs.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?view=sql-server-ver15
+LIKE_ESCAPE_TABLE = str.maketrans({char: f"\\{char}" for char in "%_[]^"})
```

Comment:
A rare sighting of `str.maketrans` in the wild!

Reply:
I was halfway through writing a hideous `re.sub` expression with three nested layers of backslash escaping when I thought to myself "there must be a better way". And there was!